### ISSUE TITLE
[0.7.2] Rx Observable?

### ISSUE THREAD

**Issue by diegoRodriguezAguila** [FabianTerhorst/Iron#4]
Any plans to have Rx Observable support?

**Comment by FabianTerhorst** [comment]
I had also the idea of implementing this, just for replacing the async tasks.

**Comment by FabianTerhorst** [comment]
I will take a look into this. I have worked a lot with rxjava and it shoudnt be to much to integrate.

**Comment by diegoRodriguezAguila** [comment]
That would be great, that way it could also be compatible with Retrofit (when using its RxJava instead of Call)

**Comment by FabianTerhorst** [comment]
@diegoRodriguezAguila thats true. I also need this to make it compatible with my [api client](https://github.com/fabianterhorst/apiclient) .

**Comment by FabianTerhorst** [comment]
I had the idea of the api that looks like this: 
```
Iron.chest().<String>get("key").subscribe(new Subscriber<String>() {

            @Override
            public void onError(Throwable e) {
                
            }

            @Override
            public void onNext(String s) {
                
            }

            @Override
            public void onCompleted() {
                
            }
        });
```

**Comment by diegoRodriguezAguila** [comment]
It looks good, it could also send updates when the related object gets modified, (instead of having to put an addOnDataChangeListener). Would Rx also be supported int the "*Store" Objects?

**Comment by FabianTerhorst** [comment]
@diegoRodriguezAguila I would add a annotation option for enabling RxJava.

**Comment by FabianTerhorst** [comment]
@diegoRodriguezAguila The idea to use the Observable for Object changes is great. Are you sure a Observable are calling the subscriber when for example a object got added to a Arraylist? 
```
ArrayList<String> strings = new ArrayList<>();
        strings.add("string");
        Observable.just(strings).subscribe(new Subscriber<ArrayList<String>>() {
            @Override
            public void onCompleted() {

            }

            @Override
            public void onError(Throwable e) {

            }

            @Override
            public void onNext(ArrayList<String> strings) {
                Log.d("array", strings.size() + "");
            }
        });
        strings.add("string");
```

Array just called once.

**Comment by diegoRodriguezAguila** [comment]
I'm not sure but I think Observable.from is for this kind of situations, in order to get notified properly, and if an item is added, tou should call on next with the item added

**Comment by FabianTerhorst** [comment]
@diegoRodriguezAguila Then the user has to use a different ArrayList like this:

```
public class ObservableList<T> {

        protected final List<T> list;
        protected final PublishSubject<T> onAdd;

        public ObservableList() {
            this.list = new ArrayList<T>();
            this.onAdd = PublishSubject.create();
        }
        public void add(T value) {
            list.add(value);
            onAdd.onNext(value);
        }
        public Observable<T> getObservable() {
            return onAdd;
        }
    }
```

**Comment by diegoRodriguezAguila** [comment]
No, that wouldn't be good, to force the user to use an specific type of collection. The key here is that as you are building the Observable and you know exactly when the event of adding the item is launched, (because is launched when the user makes something like MainStore.setContributors(contributors);) the subscriber could be saved and when the user changes the collecion, handle the event and call subscriber.onNext(...) or something like that. Or for example the Contributors list, could internally be that ObservableList wrapped object

**Comment by diegoRodriguezAguila** [comment]
I think we could get some ideas from <a href='https://github.com/pushtorefresh/storio/blob/master/storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetListOfObjects.java'>Storio's tables Observables<a>, they are also lists and I think is quite the same logic

**Comment by FabianTerhorst** [comment]
I understand you totally wrong. I realized now that you mean the change listener not the transaction. Something like this could be possible i think: Iron.chest().get("key").asObservable(new Subscriper...).

**Comment by FabianTerhorst** [comment]
Okay i decided now to go for Iron.chest().get("key"). When you add a subscriber on this onNext will get triggered on every change and on load.

**Comment by diegoRodriguezAguila** [comment]
Great! I think thats the whole point of observables, that way for example, if I have a user view, the moment the user saves its modified data, (and i'm currently subscribed on that view), I would get the data automatically updated to the screen. I think it could also work in collections? if I get all the users, (and I see the list on screen) and I add a new user the view would get updated.

**Comment by FabianTerhorst** [comment]
You can test Version 0.7.1 now. I have updated the sample app:

Iron.chest().set("rxjava", "bla");
 `...extends RxActivity`
       ```
 Iron.chest().<String>get("rxjava").compose(this.<String>bindToLifecycle()).subscribe(new Subscriber<String>() {

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onNext(String s) {
                Log.d("change", s);
            }

            @Override
            public void onCompleted() {

            }
        });

        Iron.chest().set("rxjava", "bla2");

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                Iron.chest().set("rxjava", "bla3");
            }
        }, 2000);
```

**Comment by FabianTerhorst** [comment]
The set and get methods use the scheduler threads so there are no async tasks needed anymore.

**Comment by FabianTerhorst** [comment]
I have released Version 0.7.2 now, because i found a critical bug in the callback and i also reduced the methodcount (-6k Methods) and the jar size (-600kb).

**Comment by diegoRodriguezAguila** [comment]
Awesome, I'll check it out soon. BTW I suggest you to mark the "issue" as closed and tag it with that release (0.7.2) and add a "feature" label or something like that, it's better for people searching for something similar :)
